solidity/src/Xorshift128plus.sol

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.19;
+
+/**
+ * @dev This library implements the Xorshift128+ pseudo-random number generator (PRG) algorithm.
+ */
+library Xorshift128plus {
+    /**
+     * @dev While not limited to 128 bits of state, this PRG is largely informed by xorshift128+
+     */
+    struct PRG {
+        uint64 state0;
+        uint64 state1;
+    }
+
+    /**
+     * @dev Initializer for PRG struct
+     *
+     * @param prg The PRG struct to seed
+     * @param sourceOfRandomness The entropy bytes used to seed the PRG. It is recommended to use at least 16
+     * bytes of entropy.
+     * @param salt The bytes used to salt the source of randomness
+     */
+    function seed(PRG memory prg, bytes memory sourceOfRandomness, bytes memory salt) internal pure {
+        require(
+            sourceOfRandomness.length >= 16, "At least 16 bytes of entropy should be used when initializing the PRG"
+        );
+        bytes memory tmp = abi.encodePacked(sourceOfRandomness, salt);
+        bytes32 hash = keccak256(tmp);
+
+        prg.state0 = _bigEndianBytesToUint64(abi.encodePacked(hash), 0);
+        prg.state1 = _bigEndianBytesToUint64(abi.encodePacked(hash), 8);
+
+        _requireNonZero(prg);
+    }
+
+    /**
+     * @dev Advances the PRG state and generates the next UInt64 value
+     * See https://arxiv.org/pdf/1404.0390.pdf for implementation details and reasoning for triplet selection.
+     * Note that state only advances when this function is called from a transaction. Calls from within a script
+     * will not advance state and will return the same value.
+     *
+     * @return The next UInt64 value
+     */
+    function nextUInt64(PRG memory prg) internal pure returns (uint64) {
+        _requireNonZero(prg);
```

Comment:
I guess you added this line to make sure that no one calls the function on a non-initialized `PRG` ? 

I'm asking because if we check that the state of a PRG is non zero during initialization, then it's guaranteed to never be zero later.

Reply:
This is a byproduct of how solidity libraries operate. There isn't an `init` like there is in Cadence, so one could "construct" a PRG object in memory and call `nextUInt64` before calling `seed` and since solidity fields are initialized as `0`, the state would also be `0`. This check would ensure that this implementation error could not occur or at least would do less harm by reverting than allowing for incorrect implementation with the assumption of random results.